Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,346 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Subsets as subtypes
+
+This file defines the function `set_restrict` that converts sets in a type, to sets in a subtype
+(as the elements in the subtype whose lift lives in the set), and some notation and simplification
+lemmas for convenience.
+
+More precisely, given `α : Type` and `A : Set α`, `set_restrict A : Set α → Set A` takes
+some `B : Set α` and produces some `Set ↑A` whose elements are `{ x : ↑A | ↑x ∈ B}`.
+
+It also defines the notation `A ↓∩ B` for `set_restrict A B`.
+
+## Implementation
+
+The simplification lemmas are designed to simplify the expression appearing inside the `↓∩`
+operation, if possible.
+
+## Tags
+
+subsets
+-/
+
+open Lean PrettyPrinter Delaborator SubExpr in
+/--
+Sets of a subtype coerced to the ambient type are denoted with `↑`.
+-/
+@[delab app.Set.image]
+def delab_set_image_subtype : Delab := do
+  let #[α, _, f, _] := (← getExpr).getAppArgs | failure
+  guard <| f.isAppOfArity ``Subtype.val 2
+  let some _ := α.coeTypeSet? | failure
+  let e ← withAppArg delab
+  `(↑$e)
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C: Set α} {D E : Set ↑A}
+variable {S : Set (Set α)} {T : Set (Set ↑A)} {i : β → Set α} {j : β → Set ↑A}
+
+namespace Subset
+
+/--
+Given two sets `A` and `B`, `set_restrict A B` is the set of `↑A` formed by the elements
+whose value is in `B`.
+-/
+def set_restrict (A B : Set α) : Set ↑A := restrict A B
```

Comment:
Should be camelCase per naming convention:
```suggestion
def setRestrict (A B : Set α) : Set ↑A := restrict A B
```

Reply:
Fixed